OOModel/src/CodeGenerationVisitor.cpp

```diff
@@ -0,0 +1,178 @@
+ ** OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ **
+ **********************************************************************************************************************/
+
+#include "CodeGenerationVisitor.h"
+
+#include "expressions/ReferenceExpression.h"
+#include "expressions/MetaCallExpression.h"
+
+#include "ModelBase/src/nodes/NameText.h"
+
+#include "declarations/MetaDefinition.h"
+#include "expressions/BooleanLiteral.h"
+
+namespace OOModel {
+
+CodeGenerationVisitor::CodeGenerationVisitor(QMap<QString, Model::Node *> args) : args_{args} {}
+
+void CodeGenerationVisitor::init()
+{
+	addType<ReferenceExpression>(visitReferenceExpression);
+	addType<Model::NameText>(visitNameText);
+	addType<MetaCallExpression>(visitMetaCallExpression);
+}
+
+void CodeGenerationVisitor::visitChildren(Model::Node* n)
+{
+	for (auto child : n->children()) visit(child);
+}
+
+void CodeGenerationVisitor::visitReferenceExpression(CodeGenerationVisitor* v, OOModel::ReferenceExpression* n)
+{
+	auto input = n->name();
+
+	if (!input.contains("##"))
+	{
+		if (auto argument = v->args_[input])
+		{
+			if (auto argumentReference = DCast<ReferenceExpression>(argument))
+			{
+				// case: there exists a replacement and it is another ReferenceExpression
+				// -> copy name (replacing it would cause child nodes of n to disappear)
+
+				n->setName(argumentReference->name());
+			}
+			else
+			{
+				// case: there exists a replacement and it is not a ReferenceExpression
+				// -> replace node
+
+				auto p = n->parent();
+
+				p->beginModification("change node in code generation visitor");
+
+				auto cloned = argument->clone();
+				p->replaceChild(n, cloned);
+
+				p->endModification();
+
+				// visit the cloned tree and return to avoid visiting the children of n
+				v->visitChildren(cloned);
+				return;
+			}
+		}
+	}
+	else
+	{
+		// case: n's name is a concatenated identifier (a##b)
+		// -> build identifier
+
+		QStringList parts = input.split("##");
+
+		for (auto i = 0; i < parts.size(); i++)
+			if (auto argument = DCast<ReferenceExpression>(v->args_[parts[i]]))
+				parts[i] = argument->name();
+
+		n->setName(parts.join(""));
+	}
+
+	v->visitChildren(n);
+}
+
+void CodeGenerationVisitor::visitNameText(CodeGenerationVisitor* v, Model::NameText* n)
+{
+	auto input = n->get();
+
+	if (!input.contains("##"))
+	{
+		if (auto argument = DCast<ReferenceExpression>(v->args_[input]))
+		{
+			// case: n's name is an existing argument of type ReferenceExpression
+			// -> copy name as new text
+
+			n->set(argument->name());
+		}
+	}
+	else
+	{
+		// case: n's text is a concatenated text (a##b)
+		// -> build text
+
+		QStringList parts = input.split("##");
+
+		for (auto i = 0; i < parts.size(); i++)
+			if (auto argument = DCast<ReferenceExpression>(v->args_[parts[i]]))
+				parts[i] = argument->name();
+
+		n->set(parts.join(""));
```

Comment:
consider `begin/endModifcation`

The fact that you haven't really used it and stuff seems to work can be indicative that we are only ever dealing with cases of freshly cloned nodes that do not belong to a tree manager yet. If you are sure that this will always be the case, then feel free to omit using `begin/end`. However, if you are not sure, or if you find definitive cases where we need `begin/end` use it everywhere. Regardless of the case, using `begin/end` is always safe (even if it is not needed)